bundles/org.openhab.binding.neohub/src/main/java/org/openhab/binding/neohub/internal/NeoHubHandler.java

```diff
@@ -248,23 +325,54 @@ protected NeoHubReadDcbResponse fromNeoHubReadDcbResponse() {
      * handlers
      */
     private synchronized void lazyPollingSchedulerExecute() {
-        NeoHubInfoResponse infoResponse = fromNeoHubReadInfoResponse();
-
-        if (infoResponse != null) {
-            // determine temperatureUnit
-            NeoHubReadDcbResponse dcbResponse = fromNeoHubReadDcbResponse();
-            Unit<?> temperatureUnit = (dcbResponse != null) ? dcbResponse.getTemperatureUnit() : SIUnits.CELSIUS;
+        // check which API is supported
+        if (!isApiOnline) {
+            selectApi();
+        }
 
-            // dispatch infoResponse to each of the hub's owned devices ..
+        NeoHubAbstractDeviceData deviceData = fromNeoHubGetDeviceData();
+        if (deviceData != null) {
+            // dispatch deviceData to each of the hub's owned devices ..
             List<Thing> children = getThing().getThings();
             for (Thing child : children) {
                 ThingHandler device = child.getHandler();
                 if (device instanceof NeoBaseHandler) {
-                    ((NeoBaseHandler) device).toBaseSendPollResponse(infoResponse, temperatureUnit);
+                    ((NeoBaseHandler) device).toBaseSendPollResponse(deviceData);
                 }
             }
-        }
 
+            // evaluate and update the state of our RF mesh QoS channel
+            List<? extends AbstractRecord> devices = deviceData.getDevices();
+            State state;
+
+            if (devices == null || devices.isEmpty()) {
+                state = UnDefType.UNDEF;
+            } else {
+                int totalDeviceCount = devices.size();
+                int onlineDeviceCount = 0;
+
+                for (AbstractRecord device : devices) {
+                    String deviceName = device.getDeviceName();
+                    Boolean online = !device.offline();
+
+                    if (connectionStates.containsKey(deviceName)) {
+                        @Nullable
+                        Boolean onlineBefore = connectionStates.get(deviceName);
+                        if (!online.equals(onlineBefore)) {
+                            logger.info("device \"{}\" has {} the RF mesh network", deviceName,
+                                    online.booleanValue() ? "joined" : "left");
+                        }
+                    }
+                    connectionStates.put(deviceName, online);
+
+                    if (online.booleanValue()) {
+                        onlineDeviceCount++;
+                    }
+                }
+                state = new QuantityType<>((100.0 * onlineDeviceCount) / totalDeviceCount, Units.PERCENT);
+            }
+            updateState(CHAN_MESH_NETWORK_QOS, state);
+        }
         if (fastPollingCallsToGo.get() > 0) {
             fastPollingCallsToGo.decrementAndGet();
         }
```

Comment:
`fastPollingCallsToGo` decreases every time `lazyPollingSchedulerExecute` is called. Once `fastPollingCallsToGo` reaches zero `fastPollingSchedulerExecute` will no longer call `lazyPollingSchedulerExecute`. At that point you might as well cancel the `fastPollingScheduler` since it won't do anything at that point.

Also, no point in using AtomicInteger if you aren't going to make the changes atomic.
```suggestion
        int callsToGo;
        do {
            callsToGo = fastPollingCallsToGo.get();
        } while(callsToGo > 0 && !fastPollingCallsToGo.compareAndSet(callsToGo, callsToGo - 1));
```

Reply:
> At that point you might as well cancel the fastPollingScheduler since it won't do anything at that point.

Good point. Note: as [siemensrds] uses the same code, I shall need to open yet another PR to fix this on that binding. Sorry for that :(

> Also, no point in using AtomicInteger if you aren't going to make the changes atomic.

Sorry but I don't understand what you are saying here. But I think I can use fastPollingCallsToGo-- anyway since (in Java) that is in any case an atomic operation. Or??

Reply:
> use fastPollingCallsToGo-- anyway since (in Java) that is in any case an atomic operation. Or??

To answer my own question: Nope!

Reply:
^^
Please ignore everything I have said above!
I fundamentally do **NOT** agree with your initial suggestion to cancel the thread!

The `fastPollingScheduler` is created in `initialize()` and cancelled & destroyed in `dispose()`. For the interim time between `initialize()` and `dispose()` the thread is always regularly being executed. In its execute method, if `fastPollingCallsToGo > 0` the method will actually do something, otherwise the method will simply short circuit.

So the decision whether `fastPollingScheduler` will do anything depends on the value of `fastPollingCallsToGo`  -- this value is a) decremented by one every time that `lazyPollingScheduler` executes, and b) set back to a positive value any time that the method `startFastPollingBurst()` is called. It is an interplay between the two methods: one is saying that fewer future calls are needed, whereas the other is saying that more future calls are required.

So I would like to stick with my code as designed. But if you insist to create / cancel / destroy `fastPollingScheduler` every time that it is needed, it would be a major architecture change which goes way beyond the few changes you proposed above...

Reply:
Well what you have designed works, and if you think it would be too much effort to change it then you can keep it as is.